### PR TITLE
[Messenger] Fix mssql compatibility for doctrine transport.

### LINKED ISSUE #39117: [Messenger] Doctrine transport does nothing to prevent race conditions when using sqlsrv DBAL driver
**Symfony version(s) affected**: symfony/messenger:^4.4

**Description**  
When Doctrine DBAL is set up to use SQL Server, and multiple message consumers are listening on the same doctrine transport, a message can be handled by more than one consumer simultaneously. It appears as though the DBAL platform `getWriteLockSQL()` method that the Messenger connection uses does not have compatibility with SQLServerPlatform (simply returns an empty string).

**How to reproduce**  
- Doctrine DBAL with sqlsrv connection
- Many message consumers on the same doctrine transport/bus (tested with 5)

**Possible Solution**  
Unknown if this can be fixed nor if there is a workaround. May need some documentation to explain the limitations.

**Additional context**  
Queries that are executed as captured in logs:
```
[2020-11-19 15:08:24] doctrine.DEBUG: "START TRANSACTION" [] []
[2020-11-19 15:08:24] doctrine.DEBUG: SELECT m.* FROM messenger_messages m WHERE (m.delivered_at is null OR m.delivered_at < ?) AND (m.available_at <= ?) AND (m.queue_name = ?) ORDER BY available_at ASC OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY   ["2020-11-19 14:08:24","2020-11-19 15:08:24","default"] []
[2020-11-19 15:08:24] doctrine.DEBUG: UPDATE messenger_messages SET delivered_at = ? WHERE id = ? ["2020-11-19 15:08:24","7649"] []
[2020-11-19 15:08:24] doctrine.DEBUG: "COMMIT" [] []
```

^^ Notice there is no explicit lock on the select statement

### PR DESCRIPTION
**Pull request by thsmrtone1**
Add logic for locking row for update when the doctrine dbal connection is sqlsrv. This is a quick and dirty solution, but it prevents the need to rewrite the logic due to doctrine dbal limitations.

See issue https://github.com/symfony/symfony/issues/39117

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #39117
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by derrabus** [comment]
This PR looks like it fixes the problem. Can you try to cover your changes with a test?

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 1]
```diff
@@ -161,9 +162,23 @@ public function get(): ?array
                 ->orderBy('available_at', 'ASC')
                 ->setMaxResults(1);
 
+            // use WITH (UPDLOCK, ROWLOCK) to lock table if database is SQL Server
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: I don't see what in the code below tell you are running `msqsl` or `sql server`

**Comment by thsmrtone1** [comment]
> Can you try to cover your changes with a test?

@derrabus I'm afraid this may be a bit beyond my capability.

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php (hunk 1 above)
Comment (thsmrtone1) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: `AbstractPlatform::appendLockHint($fromClause, $lockMode)` only returns the "WITH (UPDLOCK, ROWLOCK)" when the platform is an instance of SQLServerPlatform (this platform will also return empty string from `getWriteLockSQL()`

Comment (jderusse) [reply src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]: Look like `SQL Anywhere` also provides such behavior.
Id rather change the comment into `append Pessimistic lock to the FROM close`

File: src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php [hunk 2]
```diff
@@ -161,9 +162,23 @@ public function get(): ?array
...
+            $sql = $query->getSQL();
+            if (($fromPart = $query->getQueryPart('from')) &&
+                ($table = $fromPart[0]['table'] ?? null) &&
+                ($alias = $fromPart[0]['alias'] ?? null)
+            ) {
+                $fromClause = sprintf('%s %s', $table, $alias);
+                $sql = str_replace(
+                    sprintf('FROM %s WHERE', $fromClause),
+                    sprintf('FROM %s WHERE', $this->driverConnection->getDatabasePlatform()->appendLockHint($fromClause, LockMode::PESSIMISTIC_WRITE)),
+                    $query->getSQL()
```
Comment (jderusse) [inline src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php]:
```suggestion
                    $sql
```

**Comment by derrabus** [comment]
> This PR looks like it fixes the problem. Can you try to cover your changes with a test?

That's okay, I'll see if I can come up with a test.

**Review by jderusse — APPROVED** [review]

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Thank you @thsmrtone1.

**Comment by derrabus** [comment]
I've added a test in #39214.

**Pull request merged**